java/arcs/core/storage/StorageProxy.kt

```diff
@@ -217,20 +217,23 @@ class StorageProxy<Data : CrdtData, Op : CrdtOperationAtTime, T>(
     }
 
     /**
-     * Apply a CRDT operation to the [CrdtModel] that this [StorageProxy] manages, notifies read
-     * handles, and forwards the write to the [Store].
+     * Forward the CRDT operation to the [Store] that manages consistent data view, then applies
+     * the operation to the [CrdtModel] that this [StorageProxy] manages and notifies read handles
+     * as soon as the operation was applied successfully at the [Store].
      */
     suspend fun applyOp(op: Op): Boolean {
```

Comment:
I don't think this change is what we want, if the op fails to apply because of version problems, the proxy and store should exchange models to solve it, the handle should not have to wait for that.
the current code is consistent with the typescript version.

Reply:
The ts version does not cover any bit of server side crash which happens at different process. How come we allow the client's onXxx() gets called even if the server side cannot apply the model change / operation due to crash and/or failure of merging model?

Reply:
A server crash is no different from any other type of network failure (network split, propagation delays, dropped packets, ...). The local CRDT node can and should continue on the basis that its data is correct. It may be that some time later (e.g. when the server restarts) the two models are now out of sync and standard CRDT protocols need to be used to resolve that - which may mean the local node responding to an update with completely new data. However, exactly the same scenario can occur by having another CRDT node write a different model after the local write.

Note that when a server restarts it needs to bootstrap its model. If this is the persistence layer, then it needs to be aware that other writers may have advanced the state of the model and resync accordingly.

Reply:
Say a Collection handle, a client wrote / appended entities to the handle couple times but the service side kept crashing. Then the client finished / stopped its Arc and destroyed everything (storage proxy, local store, etc). The client thought it has done everything correctly but didn't know the server side failed to write all updates to backing store.

If this is intended, could we document it so Arcs developers / clients fully know this could happen to some degree. Otherwise, we may get the question: I've written a series of data w/o errors, but why I missed all of them next time I restarted the Arc w/ all handled synced.

Reply:
Btw, I completely agreed with what you said and not worried about any *unsynced* state like temporary advancing model on multiple writers, etc. I believe crdt protocols can handle all of these correctly as long as all clients and server are active/alive without seeing any troubles.

What I'm worried about is the situation where model updates cannot be eventually observed and/or broadcasted to other clients due to server crash, server termination, and/or failures of applying models like oom, etc.

Reply:
So this is something we definitely need to think about. It is critical that local writes are synchronous and particles can assume success, but we also need to consider scenarios such as you've described.

My immediate thought is to establish some sort of "success pending" tracking at the proxy level where the store eventually responds on completion (but the handle remains non-blocking on writes). If it sends an error or fails to respond within some time, the local node can re-send its state to synchronize the remote node.

Then, if an arc is being shut down it can query the proxy to see if it should wait for completion first (again, probably with some timeout - data loss may not always be 100% avoidable).

Reply:
haha, I had the same opinion of a pending list.
Then I started to think the pending list may eventually become nuclear bomb particularly on memory footprint somewhat if the server kept crashing. Please notice we do have memory constrains to pass system health review i.e. only one data copy is allowed at proxy side.

Another thought is: one-step-ahead.
A new handle write is only blocked when its previous write is not confirmed by the storage service so there is only one copy in proxy side and performance can still be sustained in most cases.

Reply:
@mykmartin I have an idea for the success-pending stuff that may work well for the use-case you identified, as well as to help with our flaky-test situation. 

I'm doing some experimentation, and after that, I'll write up a short design doc.

Tracking bug is: b/156765489

Reply:
@ianull Point taken re memory concerns. If we have a server in a crash loop, we probably need to detect that anyway regardless of the protocol for local write handling. We also wouldn't need to keep copies of each write state in the proxy; once we detect that the server is out of sync, the proxy just sends its current model (i.e. the latest "full" state).

Reply:
Yep, the point is what you mentioned: once we detect that the server is out of sync, the proxy just sends its current model. Currently the proxy doesn't actively know server-side failures / exceptions / out-of-sync, in such a case if the proxy keeps maintaining all unsynced model operations and/or updates in order to sync some time later, it would eventually cause memory issues.

Reply:
If we have a server in a crash loop, I don't think there's anything we can do to recover from that - we should shut everything down in that scenario.

There is a restart-from-X mechanism planned for clients; basically a watermark that can be compared against the server version. Perhaps that can be adapted for the scenario where the storage service crashes and doesn't come back up until the client has completed, and some kind of pending-write pause times out? If the watermark is directional (i.e. you can tell one values is earlier or later than another) then this would present as the client storing a watermark that is later than what is valid.

Bear in mind that pending writes do *not* require maintenance of a list; just a bit per proxy. It is *always* possible to recover by transmitting an entire model, and the server *will* know when it needs to ask for this. I therefore do not see any potential memory pitfalls here.